datafusion/core/benches/multi_groupby.rs

```diff
@@ -0,0 +1,131 @@
+    criterion::black_box(rt.block_on(df.collect()).unwrap());
+}
+
+fn high_cardinality_batches(
+    array_len: usize,
+    batch_size: usize,
+    schema: SchemaRef,
+) -> Vec<RecordBatch> {
+    (0..array_len / batch_size)
+        .map(|i| {
+            let data1 = (0..batch_size)
+                .map(|j| (batch_size * i + j) as i64)
+                .collect::<Vec<_>>();
+            let data2 = (0..batch_size)
+                .map(|j| format!("a{}", (batch_size * i + j)))
+                .collect::<Vec<_>>();
+
+            RecordBatch::try_new(
+                schema.clone(),
+                vec![
+                    Arc::new(Int64Array::from(data1)),
+                    Arc::new(StringArray::from(data2)),
+                ],
+            )
+            .unwrap()
+        })
+        .collect::<Vec<_>>()
+}
+
+fn low_cardinality_batches(
+    array_len: usize,
+    batch_size: usize,
+    schema: SchemaRef,
+) -> Vec<RecordBatch> {
+    (0..array_len / batch_size)
+        .map(|i| {
+            let data1 = (0..batch_size)
+                .map(|j| ((batch_size * i + j) % 4 > 1) as i64)
+                .collect::<Vec<_>>();
+            let data2 = (0..batch_size)
+                .map(|j| format!("a{}", ((batch_size * i + j) % 2)))
+                .collect::<Vec<_>>();
+
+            RecordBatch::try_new(
+                schema.clone(),
+                vec![
+                    Arc::new(Int64Array::from(data1)),
+                    Arc::new(StringArray::from(data2)),
+                ],
+            )
+            .unwrap()
+        })
+        .collect::<Vec<_>>()
+}
+
+fn create_context(
+    batches: Vec<RecordBatch>,
+    schema: SchemaRef,
+) -> Result<SessionContext> {
+    let ctx = SessionContext::new();
+
+    // declare a table in memory. In spark API, this corresponds to createDataFrame(...).
+    let provider = MemTable::try_new(schema, vec![batches])?;
+    ctx.register_table("t", Arc::new(provider))?;
+
+    Ok(ctx)
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+    let array_len = 2000000; // 2M rows
+    let batch_size = 8192;
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("a", DataType::Int64, false),
+        Field::new("b", DataType::Utf8, false),
+    ]));
+
+    c.bench_function("benchmark high cardinality", |b| {
+        let batches = high_cardinality_batches(array_len, batch_size, Arc::clone(&schema));
+        let mut ctx = create_context(batches, Arc::clone(&schema)).unwrap();
+        b.iter(|| block_on(query(&mut ctx, "select a, b, count(*) from t group by a, b order by count(*) desc limit 10")))
+    });
+
+    c.bench_function("benchmark low cardinality", |b| {
+        let batches = low_cardinality_batches(array_len, batch_size, Arc::clone(&schema));
+        let mut ctx = create_context(batches, Arc::clone(&schema)).unwrap();
+        b.iter(|| block_on(query(&mut ctx, "select a, b, count(*) from t group by a, b order by count(*) desc limit 10")))
+    });
+}
```

Comment:
```
Gnuplot not found, using plotters backend
benchmark high cardinality
                        time:   [273.00 ms 350.24 ms 451.67 ms]
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild

benchmark low cardinality
                        time:   [15.764 ms 16.531 ms 17.145 ms]
```

Reply:
main branch
```
Gnuplot not found, using plotters backend
Benchmarking benchmark high cardinality: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 11.0s.
benchmark high cardinality
                        time:   [565.74 ms 964.40 ms 1.3864 s]
                        change: [+69.461% +175.35% +332.28%] (p = 0.01 < 0.05)
                        Performance has regressed.

benchmark low cardinality
                        time:   [10.736 ms 11.140 ms 11.577 ms]
                        change: [-35.106% -31.565% -28.118%] (p = 0.00 < 0.05)
                        Performance has improved.
```

This change slightly regress for low cardinality for huge gain for high cardinality